microproxy/protocol/tls.py

```diff
@@ -91,9 +93,76 @@ def create_src_sslcontext(cert, priv_key, alpn_callback=None,
     ssl_ctx.use_certificate(cert)
     ssl_ctx.use_privatekey(priv_key)
 
-    if alpn_callback:
+    if alpn_callback and HAS_ALPN:
         ssl_ctx.set_alpn_select_callback(alpn_callback)
     if info_callback:
         ssl_ctx.set_info_callback(info_callback)
 
     return ssl_ctx
+
+
+class ServerConnection(object):
+    def __init__(self, stream):
+        self.stream = stream
+        self.alpn_resolver = None
+        self.on_alpn = None
+
+    def start_tls(self, cert, priv_key, info_callback=None,
+                  on_alpn=None, alpn_resolver=None):
+        self.alpn_resolver = alpn_resolver
+        self.on_alpn = on_alpn
+        ssl_ctx = create_src_sslcontext(
```

Comment:
For normal case, alpn callback always need to return a select alpn with simple logic.

But for our case a proxy server, we need to do a lot of thing when alpn callback,
that need to create connection to destination, and wait for server side handshaking done with selected alpn. It's a lot of things.

The original code is hard to understand that what did alpn callback do, and where and when the alpn decide.
The code I refactor to separate this two logic, that it's simple and easy understand.
When we do handshaking with client side, the on_alpn callback that we register a callback that only do connect to destination
Where and when the alpn selected? We could see that we had a resolver that resolved the destination connection result.